server/src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -0,0 +1,400 @@
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} pools updated", pdpCount);
+
+        // pool provided products
+        ids = session.createSQLQuery("SELECT id FROM cp_pool WHERE owner_id IN (?1)")
+            .setParameterList("1", ownerIds)
+            .list();
+
+        sql = "UPDATE cp2_pool_provided_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND pool_id IN (?3)";
+
+        int pppCount = this.safeSQLUpdateWithCollection(sql, ids, updated.getUuid(), current.getUuid());
+        log.debug("{} provided products updated", pppCount);
+
+        // pool derived provided products
+        sql = "UPDATE cp2_pool_derprov_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND pool_id IN (?3)";
+
+        int pdppCount = this.safeSQLUpdateWithCollection(sql, ids, updated.getUuid(), current.getUuid());
+        log.debug("{} derived provided products updated", pdppCount);
+
+        // product certificates
+        // Looks like we don't need to do anything here, since we generate them on request. By
+        // leaving them alone, they'll be generated as needed and we save some overhead here.
+
+        return updated;
+    }
+
+    /**
+     * Removes the product references currently pointing to the specified product for the given
+     * owners.
+     * <p/></p>
+     * <strong>Warning:</strong> Hibernate does not gracefully handle situations where the data
+     * backing an entity changes via direct SQL or other outside influence. While, logically, a
+     * refresh on the entity should resolve any divergence, in many cases it does not or causes
+     * errors. As such, whenever this method is called, any active ActivationKey entities should
+     * be manually evicted from the session and re-queried to ensure they will not clobber the
+     * changes made by this method on persist, nor trigger any errors on refresh.
+     *
+     * @param entity
+     *  The product other objects are referencing
+     *
+     * @param owners
+     *  A collection of owners for which to apply the reference changes
+     */
+    public void removeOwnerProductReferences(Product entity, Collection<Owner> owners) {
+        // Impl note:
+        // We're doing this in straight SQL because direct use of the ORM would require querying all
+        // of these objects and the available HQL refuses to do any joining (implicit or otherwise),
+        // which prevents it from updating collections backed by a join table.
+        // As an added bonus, it's quicker, but we'll have to be mindful of the memory vs backend
+        // state divergence.
+
+        // Impl note:
+        // We have a restriction in removeProduct which should prevent a product from being removed
+        // from an owner if it is being used by a pool. As such, we shouldn't need to manually clean
+        // the pool tables here.
+
+        // TODO:
```

Comment:
As I understand this is not necessary. Because cp_installed_products reflects what consumer does have installed on the machine. The fact that we remove an available product from an owner doesn't change that. Or is there anything more to this besides what is mentioned in the TODO?